File: JAVA/src/linkedList/singleLinkedList/SingleLinkedList.java
package linkedList.singleLinkedList;

import linkedList.LinkedList;

public class SingleLinkedList<T> implements LinkedList<T> {
	private int size;
	private Node<T> head;
	private Node<T> tail;
	
	public SingleLinkedList() {
	}

	@Override
	public int add(T value) {
		Node<T> newNode = null;
		try {
			newNode = new Node<T>(value);
			if (this.head == null) {
				this.head = newNode;
			} else {
				if (this.tail == null) {
					this.head.next = newNode;
				} else {
					this.tail.next = newNode;
				}
				this.tail = newNode;
			}
			this.size++;
			return 0;
		} catch (Exception e) {
			return -1;
		}
	}

	@Override
	public int set(int index, T value) {
		return -1;
	}

	@Override
	public int remove(int index) {

		return -1;
	}

	@Override
	public boolean remove(Object value) {
		Node<T> node = null;
		T temp = null;
		if (this.size > 0) {
			if (this.head != null) {
				node = this.head;
				temp = node.data;
				if ((temp == null && value == null) || temp.equals(value)) {
				}
			}
		}
		return false;
	}

	@Override
	public int indexOf(T value) {
		return -1;
	}

	@Override
	public T get(int index) {
		return null;
	}

	@Override
	public int size() {
		return this.size;
	}

	@Override
	public void clear() {
		this.head = null;
		this.tail = null;
		this.size = 0;
	}
	
	public void nodePrint() {
		Node<T> node = this.head;
		StringBuffer sb = new StringBuffer(node.toString()).append("\n");
		while((node = node.next) != null) {
			sb.append(node.toString()).append("\n");
		}
		System.out.println(sb.toString());
	}
}

File: C/LinkedList/singleLinkedList.c
#include <stdlib.h>
#include <string.h>

#include "singleLinkedList.h"

void init(SingleLinkedList *list, void (*destroy)(void *data)) 
{
	list -> size = 0;
	list -> destroy = destroy;
	list -> head = NULL;
	list -> tail = NULL;

	return;
}

void destroy(SingleLinkedList *list)
{	
	void *data;
	while (size(list) > 0) {
		if (remove(list, NULL, (void **)&data) == 0 && list -> destroy != NULL) {
			list -> destroy(data);
		}
	}
	memset(list, 0, sizeof(SingleLinkedList));
	return;
}

int insert(SingleLinkedList *list, Node *node, const void **data) {
	Node *newNode;
	if ((newNode = (Node *)malloc(sizeof(Node))) == NULL) { //메모리할당
		return -1;
	}

	newNode -> data = (void *)data;
	
	if (node == NULL) {
		if (size(list) == 0) {
			list -> tail = newNode;
		}
		newNode -> next = list -> head;
		list -> head = newNode;
	} else {
		if (node -> next == NULL) {
			list -> tail = newNode;
		}
		newNode -> next = node -> next;
		node -> next = newNode;
	}
	list -> size++;
	return 0;
}

int remove(SingleLinkedList * list, Node *node, void **data) {
	Node *oldNode;
	if (list -> size == 0) {
		return -1;
	}
	if (node == NULL) {
		*data = list -> head -> data;
		oldNode = list -> head;
		list -> head = list -> head -> next;
		if (sizeof(list) == 0) {
			list -> tail = NULL;
		}
	} else {
		if (node -> next == NULL) {
			return -1;
		}
		*data = node -> next -> data;
		oldNode = node -> next;
		node -> next = node -> next -> next;
		if (node -> next == NULL) {
			list -> tail = node;
		}

	}
	free(oldNode);
	list -> size--;
	return 0;
}